WIP: Localization of the keyboard layout names

DevDef opened the pull request:
- gettextize all the keyboard layout name strings with N_() macro
- add the header for the N_() macro definition (glib/gi18n.h)
- fix a bad indent (spaces instead of tabs) on line 180

DevDef committed:
Gettextize the keyboard layout name strings

- gettextize all the keyboard layout name strings with N_() macro
- add the header for the N_() macro definition (glib/gi18n.h)
- fix a bad indent (spaces instead of tabs) on line 180

DevDef committed:
Create the po template file

Create the po directory, the POTFILES.in file containing the files to be
translated (C files in libfreerdp/locale/ directory) and generate the po
template file (freerdp.pot) with xgettext.

DevDef committed:
Add a French translation po file

DevDef committed:
Make freerdp_keyboard_get_layouts() accept a optional locale param

- set the locale from the optional parameter (WIP)
- populate the RDP_KEYBOARD_LAYOUT structure with localized keyboard
layout names

akallabeth commented:
Seems like a nice start for a translation of freerdp.
Would suggest the following changes though:
* abstract the macros, a hard dependency to glib is not desired everywhere
* locale can be set outside this function too, correct? Would move that code to an abstract translation wrapper too

antenore commented:
Thanks @akallabeth 

Would it be acceptable to include libintl.h instead of the whole glib/gi18n.h where needed?

In that case we could add a compile flag like WITH_TRANSLATIONS and define the typical gettext macros, like glib does

https://github.com/GNOME/glib/blob/master/glib/gi18n.h

When WITH_TRANSLATION=OFF , we can fake those macros to do just nothing but leaving the original strings.

```c
/**
 * file: client/common/i18n.h
*/
#ifdef WITH_TRANSLATIONS
#include <libintl.h>
#endif

#include <string.h>

#ifdef WITH_TRANSLATION
#define  _(String) gettext (String)
/* HERE THERE'S MUCH MORE JOB TODO */
#define N_(String) (String)
#else
#define _(String) (String)
#define N_(String) (String)
#endif
```
And include this i18n.h (whatever name and better implementation) file everywhere or where is needed.

Regarding the second point the gettext initialization can be done everywhere in the code on condition that it is done before to use gettext obviously.
Normally is done in the main to have a consistent locale at the beginning of the program execution.

If it's not what you are expecting I need  more information from you.


Thank you!!!

DevDef commented:
Every advice is welcome, my C skills are limited and don’t understand CMake stuff. :-/

DevDef commented:
I’ve just given commit rights to both of you. Don’t hesitate to modify the code the way you think it should be… :-)

akallabeth commented:
@DevDef I´ll try to allocate some time to have this checked and thought though, but currently have some other tasks at hand.

antenore commented:
@akallabeth when you will have time, maybe, just give me some directions and I'll try to put in place the base implementation. My problem is to know where to start and how (the right way to do it).
You are already extremely busy with tons of tasks ;-)

akallabeth commented:
@antenore would need some wrapper as we can not use one single translation routine for every use case. (e.g. for you glib might be the translator of choice, but on e.g. the mac, android, ios and windows ports this would not be available)
so a simple `translator_init()` and maybe a `translator_update()` function (to set up a specific translation or change it) along with some translator macros in `winpr` would be the thing that comes to mind (so no direct inclusion of the actual translator library in the code except where it is defined)

DevDef commented:
I think that libfreerdp library shoud have its own translations and so its own po directory and gettext domain, separated from the freerdp client. For example, Remmina relies on libfreerdp, but doesn’t need the freerdp client.
So I will move the po directory inside libfreerdp. Another po directory and set of po files will have to be created if someone wants to translate the freerdp client (xfreerdp).

DevDef committed:
Set libfreerdp to have its own translation files

- move po directory into libfreerdp directory
- ajust path references inside po files comments
- set textdomain to libfreerdp instead of freerdp
